ledger/trie/trie.go

```diff
@@ -0,0 +1,345 @@
+	return t.root
+}
+
+func (t *Trie) RootHash() ledger.RootHash {
+	if t.root == nil {
+		return ledger.RootHash(hash.DummyHash)
+	}
+
+	return ledger.RootHash(t.root.Hash())
+}
+
+// TODO: Add method to add multiple paths and payloads at once and parallelize insertions that do not conflict.
+
+func (t *Trie) Insert(path ledger.Path, payload *ledger.Payload) {
+
+	current := &t.root
+	depth := uint16(0)
+	for {
+		switch node := (*current).(type) {
+		case *Branch:
+			// Since the new leaf is on this path, this branch is now dirty and needs its hash recomputed if it
+			// was already computed.
+			node.FlagDirty()
+
+			// If the key bit at the index i is a 0, move on to the left child,
+			// otherwise the right child.
+			if bitutils.Bit(path[:], int(depth)) == 0 {
+				current = &node.lChild
+			} else {
+				current = &node.rChild
+			}
+			depth++
+
+		case *Extension:
+			// Since the new leaf is on this path, this extension is now dirty and needs its hash recomputed if it
+			// was already computed.
+			node.FlagDirty()
+
+			matched := commonBits(node.path, path)
+
+			if matched == nodeHeight(node.skip)-1 {
+				// The new leaf needs to be inserted precisely at one layer before the height up to which the
+				// extension currently skips.
+				// This is a special case to avoid creating an extension node that skips nothing and to use a branch
+				// instead when a new leaf's path matches with all but the last bit of the extension's skipped path.
+
+				// Create a branch to hold the original extension node's children.
+				newBranch := NewBranch(nodeHeight(matched+1), node.lChild, node.rChild)
+
+				// Create a new leaf to be the new child of the extension, along with the aforementioned branch.
+				newLeaf := NewLeaf(nodeHeight(matched+1), path, payload)
+				t.store.Save(newLeaf.Hash(), payload)
+
+				var lChild, rChild Node
+				if bitutils.Bit(path[:], int(matched)) == 0 {
+					lChild = newLeaf
+					rChild = newBranch
+				} else {
+					lChild = newBranch
+					rChild = newLeaf
+				}
+
+				if node.height-node.skip == 1 {
+					// Node only skipped one depth, so instead of moving it down under the new branch, it also needs
+					// to be replaced with a branch.
+					*current = NewBranch(nodeHeight(matched), lChild, rChild)
+				} else {
+					// The node skipped more than one depth, so we simply shorten its skip value by one.
+					*current = NewExtension(node.height, node.skip+1, node.path, lChild, rChild)
+				}
+
+				return
+			}
+
+			if matched == nodeHeight(node.height) {
+				// The new leaf needs to be inserted precisely at the layer up to which the extension currently skips.
+				// It needs to be transformed into a branch and a new extension needs to be created for the
+				// remaining path that used to be skipped over.
+
+				// Create new extension which starts lower but skips to the original height and path.
+				newExt := NewExtension(nodeHeight(matched+1), node.skip, node.path, node.lChild, node.rChild)
```

Comment:
Currently the cases you described is handled as follows:

1. We match exactly the whole extension's path except for last bit (this is what you described): we create a branch for the new inserted leaf and:
	1.1. If the previous extension skipped only 1 height, it is now replaced with a branch
	1.2. Otherwise, its skip value is trimmed by one and its previous children are placed below the new branch, which is a sibling of the new leaf

Does that make sense or was it smth else you had in mind?

Reply:
This is done here:

```go
				if node.height-node.skip == 1 {
					// Node only skipped one depth, so instead of moving it down under the new branch, it also needs
					// to be replaced with a branch.
					*current = NewBranch(nodeHeight(matched), lChild, rChild)
				} else {
					// The node skipped more than one depth, so we simply shorten its skip value by one.
					*current = NewExtension(node.height, node.skip+1, node.path, lChild, rChild)
				}
```

Reply:
Yeah, but you didn't do that in this if statement just on the before, do you know what I mean?

Reply:
It is implicitly covered by the other cases:

* If we match exactly the path except one bit
	* If the branch is only 1 of length, replace with a branch
* We match even less of the path
	* If the branch was only 1 of length, it would be impossible to match less than 1 bit of the path. This would mean that in the previous step we made a mistake and chose the wrong node.
* We match the whole path
	* In this case we don't trim anything, so it doesn't apply.

Does that make sense?

Reply:
It is implicitly covered by the other cases:

* If we match exactly the path except one bit
	* If the branch is only 1 of length, replace with a branch
* We match even less of the path
	* If the branch was only 1 of length, it would be impossible to match less than 1 bit of the path. This would mean that in the previous step we made a mistake and chose the wrong node.
* We match the whole path
	* In this case we don't trim anything, so it doesn't apply.

Does that make sense? As I said in the last status report, the condition where `matched == nodeHeight(node.height)` is erroneous and will be removed in the next commit. This currently never happens because it's an impossible condition, therefore it just does nothing 😄

Reply:
Even though the forest stuff is WIP, I pushed it in the meantime, to avoid further confusion.